### ISSUE TITLE
 invalid slice size (ArgumentError)

### ISSUE THREAD

**Issue by noclee** [0xsauby/yasuo#12]
I'm getting the following error each_slice': invalid slice size (ArgumentError). Anyone any ideas. 

Tks,

./yasuo.rb -r 10.167.31.32 -p 80,8080,443,8443 -b all

Initiating port scan
----------------------
Using nmap scan output file nmap_output_20150303105244UTC.xml

<<<Testing host - 10.167.31.32>>>
Discovered open port: 10.167.31.32:8443
http://10.167.31.32:8443 returns HTTP 200 or 401 for every requested resource. Ignoring it
./yasuo.rb:208:in `each_slice': invalid slice size (ArgumentError)
	from ./yasuo.rb:208:in `process_nmap_scan'
	from ./yasuo.rb:82:in `run'
	from ./yasuo.rb:483:in `<main>'

rvm list | grep ruby
=* ruby-1.9.2-p330 [ x86_64 ]

[root@vcenter-plugin ~]# gem list

*** LOCAL GEMS ***

bundler (1.8.3)
bundler-unload (1.0.2)
colorize (0.7.5)
domain_name (0.5.23)
executable-hooks (1.3.2)
gem-wrappers (1.2.7)
http-cookie (1.0.2)
mechanize (2.7.3)
mime-types (2.4.3)
mini_portile (0.6.2)
minitest (1.6.0)
net-http-digest_auth (1.4)
net-http-persistent (2.9.4)
nokogiri (1.6.6.2)
ntlm-http (0.1.1)
rake (0.8.7)
rdoc (2.5.8)
rprogram (0.3.2)
ruby-nmap (0.7.0)
rubygems-bundler (1.4.4)
rvm (1.11.3.9)
text-table (1.2.4)
unf (0.1.4)
unf_ext (0.0.6)
webrobots (0.1.1)

**Comment by perfectlylogical** [comment]
What seems to be happening is that yasuo only found port 8443 open but due to the behavior it saw with the server either returning a 200 or a 403 for either the thisfilecanneverexistwtf and/or thisfilecanneverexistwtf.txt requests it  did not add the IP:PORT combination to the list to continue brute forcing against. 

Since it also did not detect any other open port from that host it moved to the block where it split the URLS into different arrays for the threading purpose but since I forgot to add a check to make sure that there was at least one IP:PORT, it is trying to split the list that is not defined or null so that is why the error is happening. 

TL;DR no usable ip:port combinations found errors out because no checks where performed to make sure at least one testable ip:port was found.

Would it be possible to post the xml file to confirm that no other open port was there or you could confirm that there was no other open port?
Would it also be possible to share if there was something running on port 8443.

**Comment by noclee** [comment]
Tks for the reason for the failure. XML below, for port 8443 there is only a ssl services sitting on this port. 

       -Patrick.
cat nmap_output_20150303105244UTC.xml 
<?xml version="1.0"?>
<?xml-stylesheet href="file:///usr/share/nmap/nmap.xsl" type="text/xsl"?>

<nmaprun scanner="nmap" args="/usr/bin/nmap -sS -sV -oX nmap_output_20150303105244UTC.xml -p 80,8080,443,8443 10.167.31.32" start="1425379964" startstr="Tue Mar  3 05:52:44 2015" version="5.51" xmloutputversion="1.03">
<scaninfo type="syn" protocol="tcp" numservices="4" services="80,443,8080,8443"/>
<verbose level="0"/>
<debugging level="0"/>
<host starttime="1425379965" endtime="1425379983"><status state="up" reason="arp-response"/>
<address addr="10.167.31.32" addrtype="ipv4"/>
<address addr="00:50:56:9F:57:EA" addrtype="mac" vendor="VMware"/>
<hostnames>
</hostnames>
<ports><port protocol="tcp" portid="80"><state state="closed" reason="reset" reason_ttl="64"/><service name="http" method="table" conf="3"/></port>
<port protocol="tcp" portid="443"><state state="closed" reason="reset" reason_ttl="64"/><service name="https" method="table" conf="3"/></port>
<port protocol="tcp" portid="8080"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="nagios-nsca" product="Nagios NSCA" method="probed" conf="10"/></port>
<port protocol="tcp" portid="8443"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="http" product="Apache Tomcat/Coyote JSP engine" version="1.1" tunnel="ssl" method="probed" conf="10"/></port>
</ports>
<times srtt="382" rttvar="1724" to="100000"/>
</host>
<runstats><finished time="1425379983" timestr="Tue Mar  3 05:53:03 2015" elapsed="18.37" summary="Nmap done at Tue Mar  3 05:53:03 2015; 1 IP address (1 host up) scanned in 18.37 seconds" exit="success"/><hosts up="1" down="0" total="1"/>
</runstats>
</nmaprun>

**Comment by noclee** [comment]
Full xml didn't make into last post, reposting. 
cat nmap_output_20150303105244UTC.xml 
<?xml version="1.0"?>
<?xml-stylesheet href="file:///usr/share/nmap/nmap.xsl" type="text/xsl"?>

<nmaprun scanner="nmap" args="/usr/bin/nmap -sS -sV -oX nmap_output_20150303105244UTC.xml -p 80,8080,443,8443 10.167.31.32" start="1425379964" startstr="Tue Mar  3 05:52:44 2015" version="5.51" xmloutputversion="1.03">
<scaninfo type="syn" protocol="tcp" numservices="4" services="80,443,8080,8443"/>
<verbose level="0"/>
<debugging level="0"/>
<host starttime="1425379965" endtime="1425379983"><status state="up" reason="arp-response"/>
<address addr="10.167.31.32" addrtype="ipv4"/>
<address addr="00:50:56:9F:57:EA" addrtype="mac" vendor="VMware"/>
<hostnames>
</hostnames>
<ports><port protocol="tcp" portid="80"><state state="closed" reason="reset" reason_ttl="64"/><service name="http" method="table" conf="3"/></port>
<port protocol="tcp" portid="443"><state state="closed" reason="reset" reason_ttl="64"/><service name="https" method="table" conf="3"/></port>
<port protocol="tcp" portid="8080"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="nagios-nsca" product="Nagios NSCA" method="probed" conf="10"/></port>
<port protocol="tcp" portid="8443"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="http" product="Apache Tomcat/Coyote JSP engine" version="1.1" tunnel="ssl" method="probed" conf="10"/></port>
</ports>
<times srtt="382" rttvar="1724" to="100000"/>
</host>
<runstats><finished time="1425379983" timestr="Tue Mar  3 05:53:03 2015" elapsed="18.37" summary="Nmap done at Tue Mar  3 05:53:03 2015; 1 IP address (1 host up) scanned in 18.37 seconds" exit="success"/><hosts up="1" down="0" total="1"/>
</runstats>
</nmaprun>

**Comment by noclee** [comment]
Tks for the reason for the failure. XML below, for port 8443 there is only a ssl services sitting on this port.

-Patrick
cat nmap_output_20150303105244UTC.xml 
<?xml version="1.0"?>
"<?xml-stylesheet href="file:///usr/share/nmap/nmap.xsl" type="text/xsl"?>"
"
<nmaprun scanner="nmap" args="/usr/bin/nmap -sS -sV -oX nmap_output_20150303105244UTC.xml -p 80,8080,443,8443 10.167.31.32" start="1425379964" startstr="Tue Mar  3 05:52:44 2015" version="5.51" xmloutputversion="1.03">
<scaninfo type="syn" protocol="tcp" numservices="4" services="80,443,8080,8443"/>
<verbose level="0"/>
<debugging level="0"/>
<host starttime="1425379965" endtime="1425379983"><status state="up" reason="arp-response"/>
<address addr="10.167.31.32" addrtype="ipv4"/>
<address addr="00:50:56:9F:57:EA" addrtype="mac" vendor="VMware"/>
<hostnames>
</hostnames>
<ports><port protocol="tcp" portid="80"><state state="closed" reason="reset" reason_ttl="64"/><service name="http" method="table" conf="3"/></port>
<port protocol="tcp" portid="443"><state state="closed" reason="reset" reason_ttl="64"/><service name="https" method="table" conf="3"/></port>
<port protocol="tcp" portid="8080"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="nagios-nsca" product="Nagios NSCA" method="probed" conf="10"/></port>
<port protocol="tcp" portid="8443"><state state="open" reason="syn-ack" reason_ttl="64"/><service name="http" product="Apache Tomcat/Coyote JSP engine" version="1.1" tunnel="ssl" method="probed" conf="10"/></port>
</ports>
<times srtt="382" rttvar="1724" to="100000"/>
</host>
<runstats><finished time="1425379983" timestr="Tue Mar  3 05:53:03 2015" elapsed="18.37" summary="Nmap done at Tue Mar  3 05:53:03 2015; 1 IP address (1 host up) scanned in 18.37 seconds" exit="success"/><hosts up="1" down="0" total="1"/>
</runstats>
</nmaprun>

**Comment by noclee** [comment]
Tks for the explanation, there's a ssl services sitting on port 8443
-Patrick. 
![screen shot 2015-03-03 at 13 53 37](https://cloud.githubusercontent.com/assets/11281143/6463717/ba0c10ac-c1ac-11e4-961a-e11f8c881ec1.png)

**Comment by perfectlylogical** [comment]
I see the xml in the source of the page, or if I edit the comment, thanks for the png. At the moment due to the way we do checks a server that responds to 200 or 403 for our garbage requests if we let them though it would be time consuming to continue against them since it will say the first URL it tries will likely be true and return false positives. 

That said i will fix the part that generates the error that you are seeing sometime today so that it is also more noticeable why it doesn't continue.

**Comment by noclee** [comment]
Is there a option to ignore port/ports so I can ignore 8443.

**Comment by perfectlylogical** [comment]
There is no option to ignore ports, i just submitted a pull request to fix the issue,
Based on the XML file it would just end there since the nagious service on 8080 also want get picked up by the filter we are using. 

A possible option would be to try running nmap without the -sV flag set and then pass that xml file to yasuo to  see if you get different results,

**Comment by 0xsauby** [comment]
@noclee As @perfectlylogical mentioned, Yasuo does not support port exclusion, neither does nmap afaik. The best you could do is perform port scan for specific ports. For eg.
nmap -Pn x.x.x.x -p80,8080,443,8081 --open -oX test.xml
And then feed the output file to Yasuo:
./yasuo -f test.xml -b all

**Comment by 0xsauby** [comment]
@perfectlylogical has fixed the error you were experiencing. Please download the latest code.

**Comment by noclee** [comment]
Great tks for the quick fix. I also found a solution for ignoring a port 
./yasuo.rb -r 10.167.31.32 -p 1-8442,8444-65535 -b form
